File: lib/src/options.dart
import 'dart:typed_data';

// ─── Shared utility ──────────────────────────────────────────────────────────

String formatBytes(int bytes) {
  if (bytes < 1024) return '$bytes B';
  if (bytes < 1024 * 1024) return '${(bytes / 1024).toStringAsFixed(1)} KB';
  return '${(bytes / (1024 * 1024)).toStringAsFixed(1)} MB';
}

// ─── Enums ───────────────────────────────────────────────────────────────────

/// Output image format.
enum CompressFormat {
  /// Keep the same format as input (default).
  auto(0),

  /// JPEG output — lossy, best for photos.
  jpeg(1),

  /// PNG output — lossless.
  png(2),

  /// WebP lossless — pure Rust, no C deps.
  /// Typically 25-35% smaller than PNG for graphics/screenshots.
  /// For photos, JPEG via mozjpeg is still smaller.
  webpLossless(3),

  /// WebP lossy — quality-based, often more useful than lossless for photos.
  /// Typically smaller than JPEG at equivalent visual quality.
  webpLossy(4);

  const CompressFormat(this.value);
  final int value;
}

/// JPEG chroma subsampling mode.
enum ChromaSubsampling {
  /// 4:2:0 — best compression, slight color loss (default, recommended).
  yuv420(0),

  /// 4:2:2 — balanced.
  yuv422(1),

  /// 4:4:4 — no chroma loss, larger files.
  yuv444(2);

  const ChromaSubsampling(this.value);
  final int value;
}

// ─── Options ─────────────────────────────────────────────────────────────────

/// Advanced JPEG encoding options.
///
/// These expose mozjpeg's powerful features that most Flutter packages
/// don't offer. All have sensible defaults.
class JpegOptions {
  const JpegOptions({
    this.progressive = true,
    this.chromaSubsampling = ChromaSubsampling.yuv420,
    this.trellis = true,
  });

  /// Enable progressive JPEG encoding. Produces smaller files for web.
  /// Default: `true`.
  final bool progressive;

  /// Chroma subsampling mode. Default: [ChromaSubsampling.yuv420].
  final ChromaSubsampling chromaSubsampling;

  /// Enable trellis quantization — mozjpeg's killer feature.
  /// Produces measurably smaller files at the same quality.
  /// Default: `true`.
  final bool trellis;
}

/// Advanced PNG optimization options.
class PngOptions {
  const PngOptions({
    this.optimizationLevel = 2,
  });

  /// Optimization level 0-6. Higher = slower but smaller output.
  ///
  /// - 0: No optimization (fastest)
  /// - 2: Good balance (default)
  /// - 6: Maximum compression (slowest)
  final int optimizationLevel;
}

// ─── Quality Presets ─────────────────────────────────────────────────────────

/// Built-in quality presets for common use cases.
///
/// Pass to [Ironpress.compressFile], [Ironpress.compressBytes], or
/// [Ironpress.compressBatch] via the `preset` parameter. Individual
/// parameters always take priority over the preset value.
///
/// ```dart
/// // One-liner with sensible defaults
/// final result = await Ironpress.compressFile(
///   'photo.jpg',
///   preset: CompressPreset.medium,
/// );
///
/// // Override one field while keeping the rest from the preset
/// final result = await Ironpress.compressFile(
///   'photo.jpg',
///   preset: CompressPreset.medium,
///   maxWidth: 1280, // override just this
/// );
/// ```
class CompressPreset {
  const CompressPreset._({
    required this.quality,
    required this.minQuality,
    this.maxWidth,
    this.maxHeight,
  });

  /// Social media / messaging. Small file size, fast upload.
  ///
  /// - JPEG quality: 65
  /// - Max dimension: 1280 px (long edge)
  /// - Min quality floor: 25 (aggressive target-size search)
  static const low = CompressPreset._(
    quality: 65,
    minQuality: 25,
    maxWidth: 1280,
    maxHeight: 1280,
  );

  /// General uploads / in-app photos. Good balance of size and quality.
  ///
  /// - JPEG quality: 80
  /// - Max dimension: 1920 px (long edge)
  /// - Min quality floor: 35
  static const medium = CompressPreset._(
    quality: 80,
    minQuality: 35,
    maxWidth: 1920,
    maxHeight: 1920,
  );

  /// High-quality archives or professional use. Minimal compression.
  ///
  /// - JPEG quality: 92
  /// - No resize applied
  /// - Min quality floor: 70
  static const high = CompressPreset._(
    quality: 92,
    minQuality: 70,
  );

  /// JPEG quality (0–100) for this preset.
  final int quality;

  /// Minimum quality floor used during binary-search file size targeting.
  final int minQuality;

  /// Maximum width in pixels, or `null` for no constraint.
  final int? maxWidth;

  /// Maximum height in pixels, or `null` for no constraint.
  final int? maxHeight;
}

// ─── Batch Input ─────────────────────────────────────────────────────────────

/// Input descriptor for batch compression.
class CompressInput {
  const CompressInput({this.path, this.data, this.outputPath})
      : assert(
          (path != null) != (data != null),
          'Exactly one of path or data must be provided',
        );

  /// File path to compress (mutually exclusive with [data]).
  final String? path;

  /// Raw image bytes to compress (mutually exclusive with [path]).
  final Uint8List? data;

  /// Output path. If null, result is returned as bytes.
  final String? outputPath;
}

// ─── Cancellation ────────────────────────────────────────────────────────────

/// Token for cancelling batch compression between chunks.
///
/// ```dart
/// final token = CancellationToken();
///
/// // Start batch
/// final future = Ironpress.compressBatch(
///   inputs,
///   cancellationToken: token,
/// );
///
/// // Cancel after 5 seconds
/// Future.delayed(Duration(seconds: 5), () => token.cancel());
///
/// final result = await future; // Contains partial results
/// ```
class CancellationToken {
  bool _cancelled = false;

  /// Whether cancellation has been requested.
  bool get isCancelled => _cancelled;

  /// Request cancellation. The batch will stop after the current chunk
  /// completes and return partial results.
  void cancel() => _cancelled = true;

  /// Reset the token for reuse.
  void reset() => _cancelled = false;
}

// ─── Result ──────────────────────────────────────────────────────────────────

/// Compression result with stats.
///
/// This provides transparency into what the compression engine actually did,
/// which is invaluable for logging and debugging.
class CompressResult {
  const CompressResult({
    this.data,
    required this.originalSize,
    required this.compressedSize,
    required this.width,
    required this.height,
    required this.qualityUsed,
    required this.iterations,
    required this.resizedToFit,
    this.errorCode,
    this.errorMessage,
  });

  /// Compressed image data. Null when output was written to a file
  /// via [Ironpress.compressFileToFile].
  final Uint8List? data;

  /// Original input size in bytes.
  final int originalSize;

  /// Compressed output size in bytes.
  final int compressedSize;

  /// Final image width in pixels.
  final int width;

  /// Final image height in pixels.
  final int height;

  /// Actual quality used by the encoder.
  /// May differ from requested if [maxFileSize] was set.
  final int qualityUsed;

  /// Number of compression iterations (1 if no [maxFileSize]).
  final int iterations;

  /// Whether the image was auto-resized to meet the file size target.
  final bool resizedToFit;

  /// Native error code for batch items that failed.
  /// Null for successful results.
  final int? errorCode;

  /// Native error message for batch items that failed.
  /// Null for successful results.
  final String? errorMessage;

  /// Compression ratio (e.g., 0.35 means 65% reduction).
  double get ratio =>
      originalSize > 0 ? compressedSize / originalSize : 1.0;

  /// Human-readable reduction percentage (e.g., "65.2%").
  String get reductionPercent =>
      '${((1.0 - ratio) * 100).toStringAsFixed(1)}%';

  /// Whether the compression completed successfully.
  bool get isSuccess => errorCode == null;

  /// Whether the result was a file-to-file operation (output written to disk,
  /// no in-memory data returned).
  bool get isFileOutput => data == null && isSuccess;

  @override
  String toString() {
    if (!isSuccess) {
      final code = errorCode != null ? '($errorCode)' : '';
      return 'CompressResult$code: ${errorMessage ?? 'unknown error'}';
    }

    return 'CompressResult('
        '${formatBytes(originalSize)} → ${formatBytes(compressedSize)} '
        '[$reductionPercent], '
        '${width}x$height, '
        'q$qualityUsed, '
        '${iterations}iter'
        '${resizedToFit ? ', auto-resized' : ''}'
        ')';
  }
}

/// Result of a batch compression operation.
///
/// Contains individual results for each input plus aggregate stats.
class BatchCompressResult {
  const BatchCompressResult({
    required this.results,
    required this.elapsedMs,
  });

  /// Individual compression results, one per input.
  final List<CompressResult> results;

  /// Total wall-clock time for the entire batch (milliseconds).
  /// This is measured inside Rust, so it excludes FFI overhead.
  final int elapsedMs;

  /// Number of successful compression results.
  int get successfulCount => results.where((r) => r.isSuccess).length;

  /// Number of failed compression results.
  int get failedCount => results.length - successfulCount;

  /// Whether any item failed during compression.
  bool get hasFailures => failedCount > 0;

  /// Total original size of all inputs combined.
  int get totalOriginalSize =>
      results.fold(0, (sum, r) => sum + r.originalSize);

  /// Total compressed size of all outputs combined.
  int get totalCompressedSize =>
      results.fold(0, (sum, r) => sum + r.compressedSize);

  /// Average compression ratio across all results.
  double get averageRatio =>
      totalOriginalSize > 0 ? totalCompressedSize / totalOriginalSize : 1.0;

  /// Throughput in images per second.
  double get imagesPerSecond =>
      elapsedMs > 0 ? results.length / (elapsedMs / 1000.0) : 0;

  /// Throughput in MB/s of input data processed.
  double get mbPerSecond =>
      elapsedMs > 0
          ? (totalOriginalSize / (1024 * 1024)) / (elapsedMs / 1000.0)
          : 0;

  @override
  String toString() =>
      'BatchCompressResult('
      '${results.length} images, '
      '${elapsedMs}ms, '
      '${imagesPerSecond.toStringAsFixed(1)} img/s, '
      '${mbPerSecond.toStringAsFixed(1)} MB/s, '
      '${((1.0 - averageRatio) * 100).toStringAsFixed(1)}% avg reduction'
      '${hasFailures ? ', $failedCount failed' : ''}'
      ')';
}

// ─── Probe Result ────────────────────────────────────────────────────────────

/// Detected image format from file header.
enum ImageFormat {
  jpeg(1),
  png(2),
  webp(3);

  const ImageFormat(this.value);
  final int value;

  static ImageFormat fromValue(int v) {
    switch (v) {
      case 1: return jpeg;
      case 2: return png;
      case 3: return webp;
      case 4: return webp;
      default: return jpeg;
    }
  }

  String get name {
    switch (this) {
      case jpeg: return 'JPEG';
      case png: return 'PNG';
      case webp: return 'WebP';
    }
  }
}

/// Quick image metadata read from file headers — no pixel decoding.
class ImageProbe {
  const ImageProbe({
    required this.width,
    required this.height,
    required this.format,
    required this.fileSize,
    required this.hasExif,
  });

  /// Image width in pixels.
  final int width;

  /// Image height in pixels.
  final int height;

  /// Detected image format.
  final ImageFormat format;

  /// File size in bytes.
  final int fileSize;

  /// Whether EXIF metadata is present.
  final bool hasExif;

  /// Total pixel count.
  int get pixelCount => width * height;

  /// Megapixels (e.g., 12.0 for 4000x3000).
  double get megapixels => pixelCount / 1000000.0;

  @override
  String toString() =>
      'ImageProbe(${width}x$height, ${format.name}, '
      '${formatBytes(fileSize)}, '
      '${megapixels.toStringAsFixed(1)}MP'
      '${hasExif ? ', EXIF' : ''})';
}

// ─── Benchmark Result ────────────────────────────────────────────────────────

/// Single quality point from a benchmark sweep.
class BenchmarkEntry {
  const BenchmarkEntry({
    required this.quality,
    required this.sizeBytes,
    required this.ratio,
    required this.encodeMs,
  });

  /// Quality level (0-100). 0 for PNG/WebP.
  final int quality;

  /// Compressed output size in bytes.
  final int sizeBytes;

  /// Compression ratio (e.g., 0.35 = 65% reduction).
  final double ratio;

  /// Encoding time in milliseconds.
  final int encodeMs;

  /// Human-readable size.
  String get sizeFormatted => formatBytes(sizeBytes);

  /// Human-readable reduction.
  String get reductionPercent =>
      '${((1.0 - ratio) * 100).toStringAsFixed(1)}%';

  @override
  String toString() =>
      'q$quality: $sizeFormatted ($reductionPercent, ${encodeMs}ms)';
}

/// Full benchmark result — quality sweep across multiple levels.
class BenchmarkResult {
  const BenchmarkResult({
    required this.originalSize,
    required this.width,
    required this.height,
    required this.format,
    required this.entries,
    required this.recommendedQuality,
  });

  /// Original file size in bytes.
  final int originalSize;

  /// Image width.
  final int width;

  /// Image height.
  final int height;

  /// Detected format.
  final ImageFormat format;

  /// Quality sweep entries, highest quality first.
  final List<BenchmarkEntry> entries;

  /// Recommended quality — best size/quality trade-off.
  final int recommendedQuality;

  @override
  String toString() {
    final buf = StringBuffer();
    buf.writeln(
        'Benchmark(${formatBytes(originalSize)}, ${width}x$height, ${format.name}, '
        'recommended: q$recommendedQuality)');
    for (final entry in entries) {
      final marker =
          entry.quality == recommendedQuality ? '  ← recommended' : '';
      buf.writeln('  $entry$marker');
    }
    return buf.toString().trimRight();
  }
}

File: rust/src/options.rs
/// Output format for compression.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u32)]
pub enum OutputFormat {
    /// Keep the same format as input
    Auto = 0,
    /// JPEG output
    Jpeg = 1,
    /// PNG output
    Png = 2,
    /// WebP output (lossless — smaller than PNG for graphics/screenshots)
    WebpLossless = 3,
    /// WebP output (lossy — often more useful than lossless for photos)
    WebpLossy = 4,
}

impl OutputFormat {
    pub fn from_u32(v: u32) -> Self {
        match v {
            1 => Self::Jpeg,
            2 => Self::Png,
            3 => Self::WebpLossless,
            4 => Self::WebpLossy,
            _ => Self::Auto,
        }
    }
}

/// Chroma subsampling mode for JPEG encoding.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u32)]
pub enum ChromaSubsampling {
    /// 4:2:0 — best compression, slight color blurring (default)
    Yuv420 = 0,
    /// 4:2:2 — balanced
    Yuv422 = 1,
    /// 4:4:4 — no chroma loss, larger files
    Yuv444 = 2,
}

impl ChromaSubsampling {
    pub fn from_u32(v: u32) -> Self {
        match v {
            1 => Self::Yuv422,
            2 => Self::Yuv444,
            _ => Self::Yuv420,
        }
    }
}

/// Full compression parameters, passed from Dart via FFI.
#[repr(C)]
pub struct CompressParams {
    /// JPEG quality 0-100 (default 80). Ignored for PNG.
    pub quality: u32,
    /// Maximum width constraint. 0 = no constraint.
    pub max_width: u32,
    /// Maximum height constraint. 0 = no constraint.
    pub max_height: u32,
    /// Target maximum file size in bytes. 0 = disabled.
    pub max_file_size: u32,
    /// Minimum quality floor for max_file_size binary search (default 30).
    pub min_quality: u32,
    /// Allow automatic downscale if quality alone can't hit max_file_size target.
    /// 1 = true, 0 = false.
    pub allow_resize: u32,
    /// Output format: 0=auto, 1=jpeg, 2=png.
    pub format: u32,
    /// Keep EXIF/metadata. 1 = true, 0 = false.
    pub keep_metadata: u32,
    /// JPEG progressive encoding. 1 = true (default), 0 = false.
    pub jpeg_progressive: u32,
    /// Chroma subsampling: 0=4:2:0, 1=4:2:2, 2=4:4:4.
    pub jpeg_chroma_subsampling: u32,
    /// JPEG trellis quantization (mozjpeg feature). 1 = true (default), 0 = false.
    pub jpeg_trellis: u32,
    /// PNG optimization level 0-6 (default 2).
    pub png_optimization_level: u32,
}

impl Default for CompressParams {
    fn default() -> Self {
        Self {
            quality: 80,
            max_width: 0,
            max_height: 0,
            max_file_size: 0,
            min_quality: 30,
            allow_resize: 1,
            format: 0,
            keep_metadata: 0,
            jpeg_progressive: 1,
            jpeg_chroma_subsampling: 0,
            jpeg_trellis: 1,
            png_optimization_level: 2,
        }
    }
}

/// Compression result returned to Dart via FFI.
#[repr(C)]
pub struct CompressResult {
    /// Pointer to compressed data. Caller must free with `free_compress_result`.
    pub data: *mut u8,
    /// Length of compressed data in bytes.
    pub data_len: usize,
    /// Original input size in bytes.
    pub original_size: usize,
    /// Final image width.
    pub width: u32,
    /// Final image height.
    pub height: u32,
    /// Actual quality used (may differ from requested if max_file_size was set).
    pub quality_used: u32,
    /// Number of compression iterations attempted (1 if no max_file_size).
    pub iterations: u32,
    /// Whether the image was auto-resized to meet max_file_size. 1 = yes.
    pub resized_to_fit: u32,
    /// Error code: 0 = success, non-zero = error.
    pub error_code: i32,
    /// Error message (null-terminated C string). Null if no error.
    pub error_message: *mut libc::c_char,
}

// Safety: CompressResult only contains raw pointers that are either null
// or exclusively owned by this result. We need Send for rayon parallelism.
unsafe impl Send for CompressResult {}

impl CompressResult {
    pub fn success(
        data: Vec<u8>,
        original_size: usize,
        width: u32,
        height: u32,
        quality_used: u32,
        iterations: u32,
        resized_to_fit: bool,
    ) -> Self {
        let data_len = data.len();
        let boxed = data.into_boxed_slice();
        let ptr = Box::into_raw(boxed) as *mut u8;
        Self {
            data: ptr,
            data_len,
            original_size,
            width,
            height,
            quality_used,
            iterations,
            resized_to_fit: if resized_to_fit { 1 } else { 0 },
            error_code: 0,
            error_message: std::ptr::null_mut(),
        }
    }

    pub fn error(code: i32, message: &str) -> Self {
        let c_msg = std::ffi::CString::new(message).unwrap_or_default();
        Self {
            data: std::ptr::null_mut(),
            data_len: 0,
            original_size: 0,
            width: 0,
            height: 0,
            quality_used: 0,
            iterations: 0,
            resized_to_fit: 0,
            error_code: code,
            error_message: c_msg.into_raw(),
        }
    }

    pub fn success_without_data(
        compressed_size: usize,
        original_size: usize,
        width: u32,
        height: u32,
        quality_used: u32,
        iterations: u32,
        resized_to_fit: bool,
    ) -> Self {
        Self {
            data: std::ptr::null_mut(),
            data_len: compressed_size,
            original_size,
            width,
            height,
            quality_used,
            iterations,
            resized_to_fit: if resized_to_fit { 1 } else { 0 },
            error_code: 0,
            error_message: std::ptr::null_mut(),
        }
    }
}

// ─── Batch Types ─────────────────────────────────────────────────────────────

/// Single input for batch compression, passed from Dart via FFI.
///
/// Exactly one of `file_path` or `data` must be non-null.
#[repr(C)]
pub struct BatchInput {
    /// Path to input file (null-terminated C string). Null if using `data`.
    pub file_path: *const libc::c_char,
    /// Pointer to input image bytes. Null if using `file_path`.
    pub data: *const u8,
    /// Length of `data` in bytes. 0 if using `file_path`.
    pub data_len: usize,
    /// Path to write output (null-terminated C string).
    /// Null to return compressed bytes in the result.
    pub output_path: *const libc::c_char,
}

// Safety: BatchInput contains only raw pointers for read-only access.
// The underlying data is owned by Dart and guaranteed valid for the
// duration of the FFI call.
unsafe impl Send for BatchInput {}
unsafe impl Sync for BatchInput {}

/// Result of a batch compression operation.
#[repr(C)]
pub struct BatchResult {
    /// Pointer to array of CompressResult. Caller must free with `free_batch_result`.
    pub results: *mut CompressResult,
    /// Number of results in the array.
    pub count: usize,
    /// Total wall-clock time in milliseconds for the entire batch.
    pub elapsed_ms: u64,
    /// Reserved for ABI compatibility with previously shipped binaries.
    /// Always null in current builds.
    pub completed: *mut u32,
}

// ─── Probe Types ─────────────────────────────────────────────────────────────

/// Quick metadata about an image without full decoding.
#[repr(C)]
pub struct ProbeResult {
    /// Image width in pixels.
    pub width: u32,
    /// Image height in pixels.
    pub height: u32,
    /// Detected format: 1=jpeg, 2=png, 3=webp.
    pub format: u32,
    /// File size in bytes.
    pub file_size: usize,
    /// Whether EXIF metadata is present. 1 = yes.
    pub has_exif: u32,
    /// Error code: 0 = success.
    pub error_code: i32,
    /// Error message (null-terminated C string). Null if no error.
    pub error_message: *mut libc::c_char,
}

impl ProbeResult {
    pub fn success(width: u32, height: u32, format: u32, file_size: usize, has_exif: bool) -> Self {
        Self {
            width,
            height,
            format,
            file_size,
            has_exif: if has_exif { 1 } else { 0 },
            error_code: 0,
            error_message: std::ptr::null_mut(),
        }
    }

    pub fn error(code: i32, message: &str) -> Self {
        let c_msg = std::ffi::CString::new(message).unwrap_or_default();
        Self {
            width: 0,
            height: 0,
            format: 0,
            file_size: 0,
            has_exif: 0,
            error_code: code,
            error_message: c_msg.into_raw(),
        }
    }
}

// ─── Benchmark Types ─────────────────────────────────────────────────────────

/// Single quality point in a benchmark sweep.
#[repr(C)]
pub struct BenchmarkEntry {
    /// Quality level used (0-100).
    pub quality: u32,
    /// Compressed output size in bytes.
    pub size_bytes: usize,
    /// Compression ratio (size / original, e.g. 0.35).
    pub ratio: f32,
    /// Encoding time in milliseconds.
    pub encode_ms: u32,
}

/// Result of a benchmark sweep across multiple quality levels.
#[repr(C)]
pub struct BenchmarkResult {
    /// Original file size in bytes.
    pub original_size: usize,
    /// Original width.
    pub width: u32,
    /// Original height.
    pub height: u32,
    /// Detected format: 1=jpeg, 2=png, 3=webp.
    pub format: u32,
    /// Pointer to array of BenchmarkEntry.
    pub entries: *mut BenchmarkEntry,
    /// Number of entries.
    pub entry_count: usize,
    /// Recommended quality for web (best size/quality trade-off).
    pub recommended_quality: u32,
    /// Error code: 0 = success.
    pub error_code: i32,
    /// Error message.
    pub error_message: *mut libc::c_char,
}
